Node: KV getter

alanmarkoTrilitech opened the pull request:
# Context

<!-- Why is this change required? What problem does it solve? -->

<!-- If it closes an Asana Task, please link to the task here. -->
<!-- **Related Tasks**: [Task name](Task url) -->

https://app.asana.com/0/1205770721173533/1205771213624617

# Description
This PR adds KV getter to the node, supporting 'kv list' and 'kv get' commands for reading values from the KV storage. It supports --name option to specify either account alias or address.
<!-- Describe your changes in detail. -->

<!-- If this PR has dependencies, please link them here. -->
<!-- **Dependencies**: -->

# Manually testing the PR
```
jstz deploy --name cntr < counter.js
jstz run "tezos://cntr/"
jstz kv list --name cnt
jstz kv get counter --name cnt
```

<!-- Describe how reviewers and approvers can test this PR. -->

johnyob commented on jstz_node/src/rollup.rs:
```diff
@@ -58,4 +60,43 @@ impl RollupClient {
+            let content_str = res.text().await?;
+            let content_json = serde_json::from_str::<Value>(&content_str);
+
+            match content_json {
+                Ok(serde_json::Value::Array(arr)) => {
+                    let list_of_strings: Result<Vec<String>> = arr
+                        .into_iter()
+                        .map(|item| match item {
+                            Value::String(s) => Ok(s),
+                            _ => Err(anyhow!("Non-string element found in the array")),
+                        })
+                        .collect();
+
+                    match list_of_strings {
+                        Ok(list) => Ok(Some(list)),
+                        Err(e) => Err(e),
+                    }
+                }
+                Ok(_) => Err(anyhow!(
+                    "Expected a JSON array but got a different structure"
+                )),
+                Err(e) => Err(anyhow!("Failed to parse content as JSON: {:?}", e)),
+            }
```
You're manually deserializing here. It would be best to write a rust type (e.g. `SubkeysResponse`) that derives a serializer and deserializer for this type (like `ValueResponse`)

johnyob commented on jstz_node/src/services/accounts.rs:
```diff
@@ -29,11 +31,63 @@ async fn nonce(
     Ok(HttpResponse::Ok().json(nonce))
 }
 
+#[get("/{address}/kv/{key:.*}")]
```
Naive question: What is the `.*` for?

alanmarkoTrilitech replied:
Key can be empty - .* ensures that it is matched even then.

johnyob replied:
Keys cannot be empty, I've just checked. 

For example, if you deploy:
```js
const KEY = "";

const handler = () => {
  let counter = Kv.get(KEY);
  console.log(`Counter: ${counter}`);
  if (counter === null) {
    counter = 0;
  } else {
    counter++;
  }
  Kv.set(KEY, counter);
  return new Response();
};

export default handler;
```
and then execute it, you obtain a `Err("EvalError: PathError: PathError::InvalidEmptyStep")` in the receipt.

johnyob commented on jstz_node/src/services/accounts.rs:
```diff
@@ -13,6 +14,7 @@ async fn nonce(
     rollup_client: Data<RollupClient>,
     path: Path<String>,
 ) -> Result<impl Responder> {
+    print!("nonce ");
```
```suggestion
```

alanmarkoTrilitech commented on jstz_cli/src/kv.rs:
```diff
@@ -0,0 +1,84 @@
+
+    let address = cfg.accounts.get_address_from(alias)?;
+
+    let key_string = match key {
+        Some(key) => key,
+        None => "".to_string(),
+    };
```
Yes, why not? it specifies the the subset of keys that we want to list.

johnyob commented on jstz_node/src/services/accounts.rs:
```diff
@@ -29,11 +30,63 @@ async fn nonce(
+) -> Result<impl Responder> {
+    let (address, key) = path.into_inner();
+
+    let storage_key = if key == "EMPTY" {
+        format!("/jstz_kv/{}", address)
+    } else {
+        format!("/jstz_kv/{}/{}", address, key)
+    };
```
I'm not a massive fan of this. I think it is good for listing the root keys, but it feels messy 🤔

johnyob replied:
Additionally keys like `TABLE/KEY` don't get matched by this endpoint

johnyob replied:
Perhaps doing `/kv?key={}` would be best for us and `/kv/subkeys?key={}` 🤔

johnyob commented on jstz_cli/src/kv.rs:
```diff
@@ -0,0 +1,84 @@
+
+        /// User address or alias
+        #[arg(short, long, value_name = "ALIAS|ADDRESS")]
+        from: Option<String>,
```
`from` is a confusing argument. Something like `--account` would be clearer

johnyob commented on jstz_node/src/rollup.rs:
```diff
@@ -17,11 +17,15 @@ pub struct ValueError {
 
 #[derive(Deserialize)]
 #[serde(untagged)]
+#[derive(Debug)]
```
Nit: Multiple derives

```suggestion
#[derive(Debug, Deserialize)]
#[serde(untagged)]
```

johnyob commented on jstz_node/src/services/accounts.rs:
```diff
@@ -29,11 +32,67 @@ async fn nonce(
+    query: web::Query<HashMap<String, String>>,
+) -> Result<impl Responder> {
+    let address = path.into_inner();
+    let key = query.get("key").cloned().unwrap_or_else(|| "".to_string());
+
+    let storage_key = if key == "" {
+        format!("/jstz_kv/{}", address)
+    } else {
+        format!("/jstz_kv/{}/{}", address, key)
+    };
```
Nit: instead of using a sentinel value (`""`) to indicate empty, we could use the `Some/None` returned by `query.get`

johnyob commented on jstz_node/src/services/accounts.rs:
```diff
@@ -29,11 +32,67 @@ async fn nonce(
+    query: web::Query<HashMap<String, String>>,
+) -> Result<impl Responder> {
+    let address = path.into_inner();
+    let key = query.get("key").cloned().unwrap_or_else(|| "".to_string());
+
+    let storage_key = if key == "" {
+        format!("/jstz_kv/{}", address)
+    } else {
+        format!("/jstz_kv/{}/{}", address, key)
+    };
```
Ditto

johnyob commented:
Overall almost mergable -- my main critic is the handling of the 'root' key vs normal keys is a bit messy. 
For root, I'd expect paths:
- `/accounts/<address>/kv`
- `/accounts/<address>/kv/subkeys`
instead of:
- `/accounts/<address>/kv?key=`
- `/accounts/<address>/kv/subkeys?key=`

In the client, I'd expect an enum for keys (either `Root` or `Key(String)` or simply reuse `Option<String>`). The API can then simply take `Into<Key>` (with `Option<String>` implementing `Into<Key>`)

alanmarkoTrilitech commented:
I've addressed the Nits. I don't think an enum fits there well. I've used Option<String> instead and refactored it so that there is less duplicated code.

johnyob commented:
> I've used Option instead and refactored it so that there is less duplicated code.

Sweet! I only suggested an enum bcs I thought the semantic differences between `Some/None` vs `Key/Root` wasn't clear, but I'm also happy just using an `Option`.

johnyob commented on jstz_node/src/services/accounts.rs:
```diff
@@ -29,11 +39,60 @@ async fn nonce(
+    query: web::Query<HashMap<String, String>>,
+) -> Result<impl Responder> {
+    let address = path.into_inner();
+    let key_option = query.get("key").cloned();
```
Nit: `.as_ref()` provides a `Option<&str>` which would avoid some clones
NB: Don't commit suggestion, other changes are required for this

```suggestion
    let key_option = query.get("key").as_ref();
```

johnyob commented on jstz_node/src/services/accounts.rs:
```diff
@@ -29,11 +39,60 @@ async fn nonce(
+async fn kv(
+    rollup_client: Data<RollupClient>,
+    path: Path<String>,
+    query: web::Query<HashMap<String, String>>,
```
Nit: Here I think it is clearer (and more consistent) to use a top-level import (avoiding the qualified `web`)

```suggestion
    query: Query<HashMap<String, String>>,
```

johnyob commented on jstz_node/src/services/accounts.rs:
```diff
@@ -29,11 +39,60 @@ async fn nonce(
+
+    let value = rollup_client.get_value(&storage_key).await?;
+
+    println!("value: {:?}", value);
```
```suggestion
```

johnyob reviewed (approved):
LGTM 🎉 

Just some v minor nits / suggestions
